Add the ability to sort ListOf's by SId.

luciansmith opened the pull request:
In my tests of Antimony, I round-trip stuff through SBML, and the order sometimes changes.  I need a function like this to sort the lists so I can do simple before-and-after comparisons.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ X] I have updated all documentation necessary.
- [ X] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically <!-- describe how it has been tested -->

luciansmith committed:
Add the ability to sort ListOf's by SId.

In my tests of Antimony, I round-trip stuff through SBML, and the order sometimes changes.  I need a function like this to sort the lists so I can do simple before-and-after comparisons.

luciansmith committed:
Add test.

fbergmann commented on src/sbml/ListOf.cpp:
```diff
@@ -685,6 +685,23 @@ ListOf::setExplicitlyListed(bool value)
 
+struct ListOfComparator
+{
+    // Compare 2 SBase* objects using name
```
the comment lies! ... it is comparing by 'id'

fbergmann commented on src/sbml/ListOf.h:
```diff
@@ -604,6 +604,11 @@ class LIBSBML_EXTERN ListOf : public SBase
   /** @endcond */
+
+  /** @cond doxygenLibsbmlInternal */
+  void sort();
```
ideally there would be a docstring here, detailing that this function sorts the list based on the id of the Sbase elements.

fbergmann commented on src/sbml/test/TestListOf.c:
```diff
@@ -215,6 +215,36 @@ START_TEST (test_ListOf_append)
+    fail_unless(child->getId() == "n");
+    child = ListOf_get(lo, 2);
+    fail_unless(child->getId() == "z");
+
```
just wondering, what would be the expected outcome when sorting an L2 list of rules?

luciansmith replied:
This comment led me to adjust the sorting so it now sorts by, in order:
* the 'id' attribute
* the results of 'getId'
* the name
* the metaid.

So, if the IDs on the rules are set, it sorts by those.  If not, it sorts by the 'variable', since that's returned from a 'getId' call.

fbergmann reviewed (dismissed):
the sorting functionality will certainly be useful, just some comments came to mind that would be nice to sort.

luciansmith committed:
Adjust from comments from Frank.

The sorting algorithm now potentially has four phases, sorting by:
  * getIdAttribute
  * getId
  * getName
  * getMetaId

Rules in particular might not have the same 'IdAttribute' as 'Id', but they're probably useful for sorting.

Comments/docstrings added to explain.

fbergmann committed:
- slight reformat for readability

fbergmann reviewed (commented):
the changes were fine with me, i just had an issue with formatting, which i changed and committed, requesting review from you again @skeating
